README.md
[![Build Status](https://travis-ci.com/NicolaCostantino/Blockchain-Internals.svg?branch=master)](https://travis-ci.com/NicolaCostantino/Blockchain-Internals)
[![codecov](https://codecov.io/gh/NicolaCostantino/Blockchain-Internals/branch/master/graph/badge.svg)](https://codecov.io/gh/NicolaCostantino/Blockchain-Internals)


# Blockchain-Internals
Proofs of Concept and sample implementations of a Blockchain in Python.

## Table of Contents

* **[Setup](#setup)**
  * [Python](#python)
  * [Requirements](#requirements)
* **[Usage](#usage)**
  * [Example](#usage-example)
* **[Development](#development)**
  * [Code](#development-code)
  * [Setup](#development-setup)
  * [Testing](#development-testing)
* **[Contributing](#contributing)**
* **[Author](#author)**
* **[License](#license)**
* **[Acknowledgments](#acknowledgments)**
  * [Inspiration](#inspiration)

## Setup <a name="setup"></a>

### Python <a name="python"></a>
Required version: 3.7.*  

Use [pyenv](https://github.com/pyenv/pyenv) or [pyenv-installer](https://github.com/pyenv/pyenv-installer) to easily switch between multiple versions of Python.  
[Pipenv](https://github.com/pypa/pipenv) will automatically suggest and install the right version for this project.

### Requirements <a name="requirements"></a>
Use [Pipenv](https://github.com/pypa/pipenv) to install the requirements needed
```bash
pipenv install
```
If [pyenv](https://github.com/pyenv/pyenv) is installed, [Pipenv](https://github.com/pypa/pipenv) will automatically suggest and install the right version for this project. 

## Usage <a name="usage"></a>
All the needed commands are listed as [GNU `make`](https://www.gnu.org/software/make/) target rules in the [Makefile](Makefile) file.  
Each subfolder could contain a local Makefile file, if needed.  
Run the target rule with:  
```bash
make <target_rule>
```
### Example <a name="usage-example"></a>
Example: Run Jupyter Labs
```bash
make jupyter_lab
```

## Development <a name="development"></a>

### Code <a name="development-code"></a>
The source code is hosted on [GitHub](https://github.com/NicolaCostantino/Blockchain-Internals).

### Setup <a name="development-setup"></a>
Use [PipEnv](https://github.com/pypa/pipenv) to install also the requirements for development and testing
```bash
pipenv install --dev
```

### Testing <a name="development-testing"></a>
Tests are executed using [PyTest](https://docs.pytest.org/en/latest/) and [pytest-cov](https://pytest-cov.readthedocs.io/en/latest/) plugin for coverage.

## Contributing <a name="contributing"></a>
Pull requests are welcome!  

For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.

## Author <a name="author"></a>
[Nicola Costantino](https://github.com/NicolaCostantino)  

## License <a name="license"></a>
[MIT](https://choosealicense.com/licenses/mit/) as listed in [LICENSE file](LICENSE)  
Copyright (c) 2019 Nicola Costantino

## Acknowledgments <a name="acknowledgments"></a>
### Inspiration <a name="inspiration"></a>
* ["Let's Build a Blockchain"](https://github.com/Haseeb-Qureshi/lets-build-a-blockchain), by Haseeb Qureshi
* ["Learn Blockchains by Building One"](https://github.com/Haseeb-Qureshi/lets-build-a-blockchain), by Daniel van Flymen

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2019-06-10
### Added
- Code shown in talk "Blockchain Internals"

[Unreleased]: https://github.com/NicolaCostantino/Blockchain-Internals/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/NicolaCostantino/Blockchain-Internals/releases/tag/v1.0.0

blockchains/mark_3/tests/test_app.py
import pytest
import random

from flask import g
from jsonschema import validate
from unittest import mock

from ...mark_1.tests.test_blockchain import (sample_transaction_f,
                                             w_sample_blocks_f)
from .. import app as sut_app
from .test_blockchain import sample_sut


SCHEMA_DRAFT = "http://json-schema.org/draft-07/schema#"


@pytest.fixture(scope="function")
def client(sample_sut):
    client = sut_app.app.test_client()
    g.blockchain = sample_sut()
    return client


@pytest.fixture(scope="function")
def schema_def_genesis_hash():
    return {
        "type": "string",
        "minLength": 1,
        "maxLength": 1,
        "pattern": "0",
    }


@pytest.fixture(scope="function")
def schema_def_hash():
    return {
        "type": "string",
        "minLength": 64,
        "maxLength": 64,
    }


@pytest.fixture(scope="function")
def schema_def_transaction():
    return {
        "type": "object",
        "properties": {
            "data": {
                "type": "string",
                "format": "date-time",
            },
        },
        "additionalProperties": False,
        "required": [
            "data",
        ]
    }


@pytest.fixture(scope="function")
def schema_def_block():
    return {
        "type": "object",
        "properties": {
            "index": {
                "type": "number",
                "minimum": 1,
            },
            "timestamp": {
                "type": "number",
            },
            "transactions": {
                "$ref": "#/definitions/transactions"
            },
            "previous_hash": {
                "oneOf": [
                    {
                        "$ref": "#/definitions/genesis_hash"
                    },
                    {
                        "$ref": "#/definitions/hash"
                    },
                ]
            }
        },
        "additionalProperties": False,
        "required": [
            "index", "timestamp", "transactions", "previous_hash"
        ]
    }


@pytest.fixture(scope="function")
def schema_def_chain(schema_def_transaction,
                     schema_def_genesis_hash,
                     schema_def_hash,
                     schema_def_block):
    return {
        "$schema": SCHEMA_DRAFT,
        "definitions": {
            "transaction": schema_def_transaction,
            "transactions": {
                "type": "array",
                "items": {
                    "$ref": "#/definitions/transaction"
                },
                "minItems": 1,
                "uniqueItems": True,
            },
            "genesis_hash": schema_def_genesis_hash,
            "hash": schema_def_hash,
            "block": schema_def_block,
            "blocks": {
                "type": "array",
                "items": {
                    "$ref": "#/definitions/block"
                },
                "minItems": 1,
                "uniqueItems": True,
            },
        },
        "type": "object",
        "properties": {
            "chain": {
                "$ref": "#/definitions/blocks"
            },
            "is_valid": {
                "type": "boolean"
            }
        },
        "additionalProperties": False,
        "required": [
            "chain", "is_valid",
        ]
    }


@pytest.fixture(scope="function")
def schema_def_mine(schema_def_block):
    return {
        "$schema": SCHEMA_DRAFT,
        "definitions": {
            "block": schema_def_block,
        },
        "type": "object",
        "properties": {
            "new_block": {
                "oneOf": [
                    {
                        "$ref": "#/definitions/block"
                    },
                    {
                        "type": "null",
                    }
                ]
            },
            "is_valid": {
                "type": "boolean"
            }
        },
        "additionalProperties": False,
        "required": [
            "new_block", "is_valid",
        ]
    }


@pytest.fixture(scope="function")
def schema_def_evil_action(schema_def_transaction,
                           schema_def_genesis_hash,
                           schema_def_hash,
                           schema_def_block):
    return {
        "$schema": SCHEMA_DRAFT,
        "definitions": {
            "transaction": schema_def_transaction,
            "transactions": {
                "type": "array",
                "items": {
                    "$ref": "#/definitions/transaction"
                },
                "minItems": 1,
                "uniqueItems": True,
            },
            "genesis_hash": schema_def_genesis_hash,
            "hash": schema_def_hash,
            "block": schema_def_block,
        },
        "type": "object",
        "properties": {
            "status": {
                "oneOf": [
                    {
                        "type": "string",
                        "pattern": "Snap done!",
                    },
                    {
                        "type": "string",
                        "pattern": "Nothing to do...",
                    }
                ]
            },
            "element": {
                "oneOf": [
                    {
                        "$ref": "#/definitions/block",
                    },
                    {
                        "type": "null",
                    }
                ]
            },
        },
        "additionalProperties": False,
        "required": [
            "status", "element",
        ]
    }


@pytest.fixture(scope="function")
def schema_def_evaluate_consensus(schema_def_transaction,
                                  schema_def_genesis_hash,
                                  schema_def_hash,
                                  schema_def_block):
    return {
        "$schema": SCHEMA_DRAFT,
        "definitions": {
            "transaction": schema_def_transaction,
            "transactions": {
                "type": "array",
                "items": {
                    "$ref": "#/definitions/transaction"
                },
                "minItems": 1,
                "uniqueItems": True,
            },
            "genesis_hash": schema_def_genesis_hash,
            "hash": schema_def_hash,
            "block": schema_def_block,
            "blocks": {
                "type": "array",
                "items": {
                    "$ref": "#/definitions/block"
                },
                "minItems": 1,
                "uniqueItems": True,
            },
        },
        "type": "object",
        "properties": {
            "status": {
                "oneOf": [
                    {
                        "type": "string",
                        "pattern": "Chain replaced",
                    },
                    {
                        "type": "string",
                        "pattern": "Chain not replaced - master",
                    }
                ]
            },
            "chain": {
                "$ref": "#/definitions/blocks",
            }
        },
        "additionalProperties": False,
        "required": [
            "status", "chain",
        ]
    }


# Tests

class TestEvilActionEndpoint(object):
    """
    Test the evil action endpoint
    """

    def test_on_empty_chain(self, client, schema_def_evil_action):
        # Arrange
        schema = schema_def_evil_action
        # Act
        response = client.post('/action/evil', json={})
        # Assert
        assert response.status_code == 200
        validate(instance=response.json, schema=schema)

    def test_on_not_empty_chain(self,
                                client,
                                schema_def_evil_action,
                                w_sample_blocks_f):
        # Arrange
        schema = schema_def_evil_action
        with client:
            random_integer = random.randint(1, 10)
            g.blockchain = w_sample_blocks_f(random_integer)
            # Act
            response = client.post('/action/evil', json={})
            # Assert
            assert response.status_code == 200
            validate(instance=response.json, schema=schema)


class TestMineEndpoint(object):
    """
    Test the mine endpoint
    """

    def test_on_empty_chain(self, client, schema_def_mine):
        # Arrange
        schema = schema_def_mine
        # Act
        response = client.post('/mine', json={})
        # Assert
        assert response.status_code == 201
        validate(instance=response.json, schema=schema)

    def test_on_not_valid_chain(self,
                                client,
                                w_sample_blocks_f,
                                schema_def_mine):
        # Arrange
        schema = schema_def_mine
        with client:
            random_integer = 3
            g.blockchain = w_sample_blocks_f(random_integer)
            g.blockchain._chain[1]['index'] = 23
            # Act
            response = client.post('/mine', json={})
            # Assert
            assert response.status_code == 200
            validate(instance=response.json, schema=schema)


class TestTransactionEndpoint(object):
    """
    Test the transaction endpoint
    """

    def test_sample_transaction(self, client, sample_transaction_f):
        # Arrange
        payload = sample_transaction_f()
        # Act
        response = client.post('/transactions/new', json=payload)
        # Assert
        assert response.status_code == 201

    def test_empty_transaction(self, client):
        # Arrange
        payload = {}
        # Act
        response = client.post('/transactions/new', json=payload)
        # Assert
        assert response.status_code == 400


class TestChainEndpoint(object):
    """
    Test the chain endpoint
    """

    def test_validate_schema(self, client, schema_def_chain):
        # Arrange
        schema = schema_def_chain
        # Act
        response = client.get('/chain', json={})
        # Assert
        assert response.status_code == 200
        validate(instance=response.json, schema=schema)


class TestRegisterKnownNodeFunction(object):
    """
    Test the register_known_node function
    """

    def test_empty_url_address(self):
        # Arrange
        test_url = ""
        # Act
        # Assert
        with pytest.raises(ValueError):
            sut_app.register_known_node(test_url)

    def test_netloc_url_address(self, client):
        # Arrange
        test_url = "http://localhost:8082"
        extracted_test_url = "localhost:8082"
        # Act
        # Assert
        with client:
            sut_app.register_known_node(test_url)
            assert extracted_test_url in g.known_nodes

    def test_path_url_address(self, client):
        # Arrange
        test_url = "localhost:8082"
        extracted_test_url = "localhost:8082"
        # Act
        # Assert
        with client:
            sut_app.register_known_node(test_url)
            assert extracted_test_url in g.known_nodes


class TestRegisterNodesEndpoint(object):
    """
    Test the register known nodes endpoint
    """

    def test_empty_node_list(self, client):
        # Arrange
        schema = {
            "$schema": SCHEMA_DRAFT,
            "type": "object",
            "properties": {
                'error': {
                    "type": "string",
                    "pattern": "Missing list of nodes",
                }
            },
            "additionalProperties": False,
            "required": [
                "error",
            ]
        }
        payload = {}
        # Act
        response = client.post('/nodes/register', json=payload)
        # Assert
        assert response.status_code == 400
        validate(instance=response.json, schema=schema)

    def test_mixed_invalid_node_list(self, client):
        # Arrange
        schema = {
            "$schema": SCHEMA_DRAFT,
            "type": "object",
            "properties": {
                "total_nodes": {
                    "type": "array",
                    "items": {
                        "type": "string",
                    },
                    "minItems": 0,
                    "uniqueItems": True,
                },
                'invalid_nodes': {
                    "type": "array",
                    "items": {
                        "type": "string",
                    },
                    "minItems": 0,
                    "uniqueItems": True,
                },
            },
            "additionalProperties": False,
            "required": [
                "total_nodes",
                "invalid_nodes"
            ]
        }
        payload = {
            "nodes": [
                "",
                "localhost:8082",
            ]
        }
        # Act
        response = client.post('/nodes/register', json=payload)
        # Assert
        assert response.status_code == 201
        validate(instance=response.json, schema=schema)

    def test_valid_node_list(self, client):
        # Arrange
        schema = {
            "$schema": SCHEMA_DRAFT,
            "type": "object",
            "properties": {
                "total_nodes": {
                    "type": "array",
                    "items": {
                        "type": "string",
                    },
                    "minItems": 0,
                    "uniqueItems": True,
                }
            },
            "additionalProperties": False,
            "required": [
                "total_nodes",
            ]
        }
        payload = {
            "nodes": [
                "http://localhost:8082",
                "localhost:8082",
            ]
        }
        # Act
        response = client.post('/nodes/register', json=payload)
        # Assert
        assert response.status_code == 201
        validate(instance=response.json, schema=schema)


class TestDoGossipFunction(object):
    """
    Test the do_gossip function
    """

    def test_empty_node_list(self):
        # Arrange
        known_nodes = []
        # Act
        retrieved_value = sut_app.do_gossip(known_nodes)
        # Assert
        assert retrieved_value == []

    def test_valid_node_list(self):
        # Arrange
        known_nodes = [
            "localhost:8081",
        ]
        with mock.patch('requests.get') as mock_get:
            mock_get.return_value.status_code = 200
            mock_get.return_value.json.return_value = {
                "chain": "sample value"
            }
            # Act
            retrieved_value = sut_app.do_gossip(known_nodes)
            # Assert
            assert retrieved_value == ["sample value"]

    def test_invalid_response_list(self):
        # Arrange
        known_nodes = [
            "localhost:8081",
        ]
        with mock.patch('requests.get') as mock_get:
            mock_get.return_value.status_code = 500
            # Act
            retrieved_value = sut_app.do_gossip(known_nodes)
            # Assert
            assert retrieved_value == []

    def test_exception_response_list(self):
        # Arrange
        known_nodes = [
            "localhost:8081",
        ]
        with mock.patch('requests.get') as mock_get:
            mock_get.side_effect = Exception()
            # Act
            retrieved_value = sut_app.do_gossip(known_nodes)
            # Assert
            assert retrieved_value == []


class TestConsensusEndpoint(object):
    """
    Test the consensus endpoint
    """

    def test_replaced_chain(self, client, schema_def_evaluate_consensus):
        # Arrange
        schema = schema_def_evaluate_consensus
        with client:
            sample_blockchain = g.blockchain.get_chain()
            mocked_blockchain = mock.Mock()
            mocked_blockchain.evaluate_consensus.return_value = True
            mocked_blockchain.get_chain.return_value = sample_blockchain
            g.blockchain = mocked_blockchain
            g.known_nodes = []
            with mock.patch('requests.get') as mock_get:
                mock_get.return_value.status_code = 200
                mock_get.return_value.json.return_value = sample_blockchain
                # Act
                response = client.post('/nodes/consensus', json={})
                # Assert
                assert response.status_code == 200
                validate(instance=response.json, schema=schema)

    def test_not_replaced_chain(self, client, schema_def_evaluate_consensus):
        # Arrange
        schema = schema_def_evaluate_consensus
        with client:
            sample_blockchain = g.blockchain.get_chain()
            mocked_blockchain = mock.Mock()
            mocked_blockchain.evaluate_consensus.return_value = False
            mocked_blockchain.get_chain.return_value = sample_blockchain
            g.blockchain = mocked_blockchain
            g.known_nodes = []
            with mock.patch('requests.get') as mock_get:
                mock_get.return_value.status_code = 200
                mock_get.return_value.json.return_value = sample_blockchain
                # Act
                response = client.post('/nodes/consensus', json={})
                # Assert
                assert response.status_code == 200
                validate(instance=response.json, schema=schema)


blockchains/mark_2/test_blockchain.py
import pytest
import random

from freezegun import freeze_time
from unittest import mock

from .blockchain import Blockchain as BlockchainSUT
from ..mark_1.tests import test_blockchain as parent


@pytest.fixture(scope="function")
def sample_sut():

    def _make_sample_sut(*args, **kwargs):
        return BlockchainSUT(*args, **kwargs)

    return _make_sample_sut


@pytest.fixture(scope="function")
def genesis_transaction():
    return {
        'unlock': '',
        'lock': ('output = The Times 03/Jan/2009 ' +
                 'Chancellor on brink of second bailout for banks'),
    }


@pytest.fixture(scope="function")
def initial_blockchain(sample_sut):

    def _make_initial_blockchain(sample_block_timestamp,
                                 genesis_block_data):
        with freeze_time(sample_block_timestamp):
            sut = sample_sut()
            sut._chain = []
            sut.current_transactions = [
                genesis_block_data,
            ]
            # Act
            retrieved_value = sut._create_genesis_block()
            return sut, retrieved_value

    return _make_initial_blockchain


@pytest.fixture(scope="function")
def sample_transaction_f():

    def _make_sample_transaction(unlock='', lock=''):
        return {
            'unlock': unlock if unlock else '',
            'lock': lock if lock else 'output=None',
        }

    return _make_sample_transaction


@pytest.fixture(scope="function")
def w_current_transactions_f(sample_transaction_f, sample_sut):

    def _make_with_transactions(number):
        sut = sample_sut()
        for iteration in range(number):  # pylint: disable=unused-variable
            sut.add_transaction(**sample_transaction_f())
        return sut

    return _make_with_transactions


class TestAddTransactionCreation(object):
    """
    Test the creation of a add transaction in a block
    """

    def test_on_empty_block_transactions_length(self,
                                                sample_sut,
                                                sample_transaction_f):
        # Arrange
        sut = sample_sut()
        # Act
        sample_transaction = sample_transaction_f()
        sut.add_transaction(**sample_transaction)
        # Assert
        assert len(sut.current_transactions) == 1

    def test_on_empty_block_transactions_value(self,
                                               sample_sut,
                                               sample_transaction_f):
        # Arrange
        sut = sample_sut()
        # Act
        sample_transaction = sample_transaction_f()
        sut.add_transaction(**sample_transaction)
        expected_transaction = {
            'input': sample_transaction['unlock'],
            'script': sample_transaction['lock'],
            'output': None,
        }
        # Assert
        assert sut.current_transactions[0] == expected_transaction

    def test_on_empty_block_custom_transactions_value(self,
                                                      sample_sut,
                                                      sample_transaction_f):
        # Arrange
        sut = sample_sut()
        # Act
        sample_transaction = sample_transaction_f(unlock='a = 1',
                                                  lock='output = a')
        sut.add_transaction(**sample_transaction)
        expected_transaction = {
            'input': sample_transaction['unlock'],
            'script': sample_transaction['lock'],
            'output': 1,
        }
        # Assert
        assert sut.current_transactions[0] == expected_transaction

    def test_on_not_empty_block_transactions_length(self,
                                                    w_current_transactions_f,
                                                    sample_transaction_f):
        # Arrange
        current_transaction_number = random.randint(1, 10)
        sut = w_current_transactions_f(current_transaction_number)
        prev_len = len(sut.current_transactions)
        # Act
        sample_transaction = sample_transaction_f()
        sut.add_transaction(**sample_transaction)
        # Assert
        assert len(sut.current_transactions) == prev_len + 1

    def test_on_not_empty_block_transactions_value(self,
                                                   w_current_transactions_f,
                                                   sample_transaction_f):
        # Arrange
        current_transaction_number = random.randint(1, 10)
        sut = w_current_transactions_f(current_transaction_number)
        # Act
        sample_transaction = sample_transaction_f()
        sut.add_transaction(**sample_transaction)
        expected_transaction = {
            'input': sample_transaction['unlock'],
            'script': sample_transaction['lock'],
            'output': None,
        }
        # Assert
        assert sut.current_transactions[-1] == expected_transaction


class TestCreateGenesisBlockMethod(object):
    """
    Test the create genesis block method
    """

    def test_add_transaction_called(self, sample_sut):
        # Arrange
        sut = sample_sut()
        sut._chain = []
        with mock.patch.object(sut, 'add_transaction'):
            # Act
            sut._create_genesis_block()
            # Assert
            sut.add_transaction.assert_called()  # pylint: disable=E1101

    def test_return_the_genesis_block(self,
                                      genesis_transaction,
                                      initial_blockchain):
        # Arrange
        sample_block_timestamp = '2009-01-03 18:15:05'
        genesis_block_data = genesis_transaction
        # Act
        sut, genesis_block = initial_blockchain(sample_block_timestamp,
                                                genesis_block_data)
        # Assert
        assert sut.genesis_block == genesis_block

    def test_return_None_if_already_started(self, sample_sut):
        # Arrange
        sut = sample_sut()
        # Act
        retrieved_value = sut._create_genesis_block()
        # Assert
        assert retrieved_value == None


blockchains/mark_3/blockchain.py
from ..mark_1.blockchain import Blockchain as BlockchainMark1
from ..mark_1 import validation


class Blockchain(BlockchainMark1):
    def is_valid(self, chain=None):
        """
        Determine the blockchain is valid

        :return: True if valid, False if not
        """
        chain_to_validate = chain if chain else self._chain
        return validation.is_a_valid_chain(chain_to_validate)

    def evaluate_consensus(self, collected_chains):
        """
        The consensus algorithm.
        The internal chain is replaced with the received one if longer.

        :return: True if our chain was replaced, False if not
        """
        new_chain = None
        max_length = len(self.get_chain())

        for chain in collected_chains:
            length = len(chain)
            if length > max_length and self.is_valid(chain):
                # Found a new master blockchain
                max_length = length
                new_chain = chain

        if new_chain:
            self._chain = new_chain
            return True
        return False


blockchains/mark_3/app.py
import os
import requests

from flask import Flask, g, jsonify, request
from flask.views import View
from urllib.parse import urlparse
from uuid import uuid4

from .blockchain import Blockchain


# Instantiate the Node
app = Flask(__name__)
ctx = app.app_context()
ctx.push()

# Instantiate the Blockchain
g.blockchain = Blockchain()

# Generate a globally unique address for this node
g.node_identifier = str(uuid4()).replace('-', '')

# List of neighbours to this node
g.known_nodes = set()


# Blockchain-related actions

@app.route('/action/evil', methods=['POST'])
def evil_action_pop():
    blockchain = g.blockchain
    if len(blockchain._chain) > 1:
        # Don't remove a genesis block
        removed_element = blockchain._chain.pop()
        result = {
            'status': 'Snap done!',
            'element': removed_element,
        }
        status_code = 200
    else:
        result = {
            'status': 'Nothing to do...',
            'element': None,
        }
        status_code = 200
    response = jsonify(result)
    response.status_code = status_code
    return response


@app.route('/mine', methods=['POST'])
def mine():
    blockchain = g.blockchain
    result = {
        'new_block': blockchain.mine(),
        'is_valid': blockchain.is_valid(),
    }
    response = jsonify(result)
    status_code = 200 if not result['is_valid'] else 201

    response = jsonify(result)
    response.status_code = status_code
    return response


@app.route('/transactions/new', methods=['POST'])
def new_transaction():
    blockchain = g.blockchain
    input_values = request.get_json()

    required_args = ['data', ]
    if not input_values or not all(k in input_values for k in required_args):
        result = {
            'error': 'Missing values',
        }
        status_code = 400
    else:
        transaction = blockchain.add_transaction(input_values['data'])
        result = {
            'transaction': transaction,
        }
        status_code = 201

    response = jsonify(result)
    response.status_code = status_code
    return response


@app.route('/chain', methods=['GET'])
def chain():
    blockchain = g.blockchain
    result = {
        'chain': blockchain.get_chain(),
        'is_valid': blockchain.is_valid(),
    }

    response = jsonify(result)
    response.status_code = 200
    return response


# Node-related actions

def register_known_node(url_address):
    """
    Add a new node to the list of known nodes

    :param url_address: Address of node. Eg. 'http://127.0.0.1:5000'
    """
    # Parse the url
    parsed_url = urlparse(url_address)

    if parsed_url.netloc:
        # URL with scheme, e.g. 'http://127.0.0.1:5000'.
        extracted_url = parsed_url.netloc
        g.known_nodes.add(extracted_url)
    elif parsed_url.path:
        # URL without scheme, e.g. '127.0.0.1:5000'.
        extracted_url = parsed_url.path
        g.known_nodes.add(extracted_url)
    else:
        raise ValueError('Invalid URL')


@app.route('/nodes/register', methods=['POST'])
def register_nodes():
    input_values = request.get_json()

    nodes = input_values.get('nodes', None)
    if not nodes:
        result = {
            'error': 'Missing list of nodes',
        }
        status_code = 400
    else:
        invalid_nodes = []
        for node in nodes:
            try:
                register_known_node(node)
            except ValueError:
                print(f'Error adding {node}')
                invalid_nodes.append(node)
        result = {
            'total_nodes': list(g.known_nodes),
        }
        if len(invalid_nodes):
            result['invalid_nodes'] = invalid_nodes
        status_code = 201

    response = jsonify(result)
    response.status_code = status_code
    return response


def do_gossip(known_nodes):
    """
    The gossip algorithm - pull.
    All the known nodes are checked for collecting the available chains.

    :return: True if our chain was replaced, False if not
    """
    collected_chains = []
    # Pull: Collect the blockchains from the neighbours
    for node in known_nodes:
        try:
            response = requests.get(f'http://{node}/chain')

            if response.status_code == 200:
                chain = response.json()['chain']
                collected_chains.append(chain)
        except Exception as e:
            print(f'Exception on node {node}: {e}')
    return collected_chains


@app.route('/nodes/consensus', methods=['POST'])
def evaluate_consensus():
    """
    The consensus algorithm.
    The internal chain is replaced with the received one if longerk.

    :return: True if our chain was replaced, False if not
    """
    blockchain = g.blockchain
    collected_chains = do_gossip(g.known_nodes)
    replaced = blockchain.evaluate_consensus(collected_chains)

    if replaced:
        result = {
            'status': 'Chain replaced',
            'chain': blockchain.get_chain(),
        }
    else:
        result = {
            'status': 'Chain not replaced - master',
            'chain': blockchain.get_chain(),
        }
    status_code = 200

    response = jsonify(result)
    response.status_code = status_code
    return response
